Url search

afinn12 opened the pull request:
## Needed
- Merge PR [#20](https://github.com/kata-containers/kata-containers.github.io/pull/20) before this 

## Features
- Implemented form that handles searching.
  - Appends search terms to URL.
    - matchMode= matchAny/matchAll (or/and)
    - Filters job based on "value" --> contains only
  - "Keep Search Terms" button keeps the existing search terms in the URL, adding the new rule.
    - If not selected, the search erases the previous search terms, only searching for the new rule.
  - "Clear Search" button clears the search terms, if they exist.

## Testing
- [x] Tested in dev
- [x] Tested in prod

**in prod:**

 Default:
![image](https://github.com/user-attachments/assets/a832060b-d87c-49ef-a562-296e45fff8c1)

Searching with the form appends to URL:
![image](https://github.com/user-attachments/assets/de468898-24ce-4ae7-9980-6496621c4809)

Adding another search with "Keep Search Terms" appends to the URL:
(will update matchMode (and/or) if it changes)
![image](https://github.com/user-attachments/assets/297056d5-626d-429c-bf26-1b32aefedd28)


Adding another search *without* "Keep Search Terms" appends to the URL:
![image](https://github.com/user-attachments/assets/2cff64cf-4236-4181-af3b-fc45cea8e727)

Selecting "Clear Search":
![image](https://github.com/user-attachments/assets/5e003945-482c-413b-aa1d-7d21b584ce08)

a1icja commented on pages/index.js:
```diff
@@ -1,21 +1,24 @@
 
       if (process.env.NODE_ENV === "development") {
-        data = (await import("../job_stats.json")).default;
+        data = (await import("../data/job_stats.json")).default;
```
Please leave alone as per #20 

```suggestion
        data = (await import("../job_stats.json")).default;
```

a1icja commented on pages/index.js:
```diff
@@ -1,21 +1,24 @@
   const [jobs, setJobs] = useState([]);
   const [rows, setRows] = useState([]);
   const [expandedRows, setExpandedRows] = useState([]);
+  const [keepSearch,     setKeepSearch]     = useState(true);
```
```suggestion
  const [keepSearch, setKeepSearch] = useState(true);
```

a1icja commented on pages/index.js:
```diff
@@ -41,15 +44,55 @@ export default function Home() {
     fetchData();
   }, []);
 
+  // Filters the jobs s.t. all values must be contained in the name.
+  const matchAll = (filteredJobs, urlParams) => {
+    const values = urlParams.getAll("value");
+    return filteredJobs.filter((job) => {
+        const jobName = job.name.toLowerCase();
+        return values.every((val) => {
+            const decodedValue = decodeURIComponent(val).toLowerCase();
+            return jobName.includes(decodedValue);
+        });
+    });
+  };
+  
+  // Filters the jobs s.t. at least one value must be contained in the name.
+  const matchAny = (filteredJobs, urlParams) => {
+    const values = urlParams.getAll("value");
+    return filteredJobs.filter((job) => {
+        const jobName = job.name.toLowerCase();
+        return values.some((val) => {
+            const decodedValue = decodeURIComponent(val).toLowerCase();
+            return jobName.includes(decodedValue); 
+        });
+    });
+  };
+  
+
```
Is there not already an existing implementation for this somewhere? Just curious

a1icja commented on pages/index.js:
```diff
@@ -41,15 +44,55 @@ export default function Home() {
+
+    //Filter based on the URL. 
+    const urlParams = new URLSearchParams(window.location.search);
+    if(urlParams.get("matchMode") === "and"){
+      filteredJobs = matchAll(filteredJobs, urlParams);
+    }else if(urlParams.get("matchMode") === "or"){
+      filteredJobs = matchAny(filteredJobs, urlParams);
+    }
```
nit: I'm partial to switch statements in cases like this due to readability

a1icja commented on pages/index.js:
```diff
@@ -178,9 +259,46 @@ export default function Home() {
+        </div>
+        
+        <div className="mt-1 text-center md:text-lg text-base">
+          Total Rows: {rows.length}
```
Could this be moved into its own component?

afinn12 replied:
Will do in small enhancements

a1icja commented on pages/index.js:
```diff
@@ -1,21 +1,24 @@
+
```
```suggestion

```

a1icja reviewed (changes requested):
Minor changes again - Sorry D:

afinn12 commented:
## Changes
- Removed Clear Search/Keep Search Terms
- Removed matchMode (any/all) 
  - Only supports 1 term at a time

To reimplement, define clearSearch:
```  
// Clear the search parameters, but only if they exist.
  const clearSearch = () => {
    if(window.location.href.includes("matchMode")){
      const path = new URLSearchParams();
      path.append("display", display);
      if (display === "prsingle" && selectedPR) {
        path.append("pr", selectedPR);
      }
      window.location.assign(`${basePath}/?${path.toString()}`);
    }
  };
```
Modify handleSearch:
```
const handleSearch= (e) => {
    // Prevent the default behavior so that we can keep search terms.
    e.preventDefault();
    const matchMode = e.target.matchMode.value;
    const value = e.target.value.value.trimEnd(); 
    if (value) {  
      // Append the new matchMode regardless of if search terms were kept.
      const path = new URLSearchParams();
      path.append("display", display);
      if(display === "prsingle" && selectedPR){
        path.append("pr", selectedPR);
      }
      path.append("matchMode", matchMode);
      if (keepSearch) {
        // If keepSearch is true, add existing parameters in the URL.
        const urlParams = new URLSearchParams(window.location.search);
        urlParams.getAll("value").forEach((val) => {
          path.append("value", val); 
        });
      }
      //Add the search term from the form and redirect. 
      path.append("value", encodeURIComponent(value)); 
      window.location.assign(`${basePath}/?${path.toString()}`);
    }
  };
```
Add buttons to use:
```
      <button 
              className={buttonClass()} 
              onClick={() => clearSearch()}>
              Clear Search
            </button>
            <button 
              className={buttonClass(keepSearch)} 
              onClick={() => setKeepSearch(!keepSearch)}>
              Keep URL Search Terms
       </button>

```
- Users don't have to manually clear the URL
